chain/boltdb/store.go

```diff
@@ -52,19 +55,27 @@ func NewBoltStore(l log.Logger, folder string, opts *bolt.Options) (*BoltStore,
 	}, err
 }
 
-// Len performs a big scan over the bucket and is _very_ slow - use sparingly!
-func (b *BoltStore) Len(context.Context) (int, error) {
-	var length = 0
-	err := b.db.View(func(tx *bolt.Tx) error {
-		bucket := tx.Bucket(beaconBucket)
-		// this `.Stats()` call is the particularly expensive one!
-		length = bucket.Stats().KeyN
-		return nil
-	})
+// Len uses the heuristic of finding the current round and adding one to guess the length of the store.
+// This was due to limitations in bolt store which make the len call in stats being veeeeery slow
+func (b *BoltStore) Len(c context.Context) (int, error) {
+	lastBeacon, err := b.Last(c)
```

Comment:
Would this work with the beacon pruning feature?
What if we take a different approach and do the following:
- In the `NewBoltStore` we compute the "proper" BoltDB len using the Stats call.
- We then store that in a `BoltStore.size` variable.
- Whenever we invoke `Add` or `Del` we lock the BoltStore and then increment/decrement the `BoltStore.size`.
- In the `Len` call, we lock/unlock the BoltStore and return the `BoltStore.size` value.

Reply:
>Would this work with the beacon pruning feature?
definitely not :p

>What if we take a different approach
I like this approach, but if it takes up to several minutes for a big table scan that will be quite noticeable at startup. Definitely better to amortise it at startup than do it on every `Len()` call though